Switch to URL path-based identifiers.

peichman-umd opened the pull request:
This is a very rough first draft of switching from the header-based transaction identifiers to scoping transactions via URL paths. This approach is similar to the way the Fedora 4 and 5 Java Modeshape-based implementations work.

peichman-umd committed:
Switch to URL path-based identifiers.

bbpennel commented:
Using a path segment to identify the transaction the client wants to interact with potentially places a lot more burden on the client. It needs to understand the the URL it creates is not the permanent URL of the resource, and either the client or fedora needs to be able to resolve RDF references to resources within a transaction back to their non-transaction URL at commit time. 

I'm still thinking through the value of path based transactions, but I don't think it requires path based transaction identifiers. A particular transaction id could still only apply to a particular path within fedora as a header. My experience with the fedora 4/5 transaction model and cleaning up identifiers in transactions makes me inclined towards a header approach.

peichman-umd commented:
@bbpennel I definitely want to omit the transaction identifier from the URIs in any graphs returned while in a transaction. I agree that on the client side wrangling with those dual URIs is a pain (see for instance what we do in our batch loading client: https://github.com/umd-lib/plastron/blob/2.2.0/plastron/pcdm.py#L201-L215)

I'll take a crack at moving the transaction identifier into the headers.

peichman-umd committed:
Use a mixture of headers and URIs.

Keeps the transaction ID out of the resource URIs. But make the transaction ID a
URI so it can be acted upon for commit/rollback/keepalive operations.

peichman-umd commented:
@bbpennel I've updated to remove the txn ID from the resource URIs. I'm retaining a transaction URI as the `Atomic-ID` header and the URI for manipulating the transaction itself.

peichman-umd committed:
Use Trusty OS version to fix Travis build error.

whikloj commented:
Just wondering about this:
> If the client sending the current request is not authorized to access the transaction identified by the <code>Atomic-ID</code> header, the server MUST NOT update any resources, and MUST respond with a <code>404 Not Found</code> HTTP status.

What does it mean to be _not authorized_ to access a transaction?

peichman-umd commented:
@whikloj I think that is a holdover in my mind from Fedora 4, where (I believe) there was a notion that only the client who initiated a transaction could update it. I'm fine with striking that, and going to a model of using the transaction URI as the only "credential" needed to access the transaction.

If so, I might add some guidance that transaction URIs should not be easily guessable...

bbpennel commented on index.html:
```diff
@@ -93,80 +98,172 @@ <h2>Atomic Operations</h2>
+
+          201 Created
+          Location: http://example.org/foo/abcdef01-2345-6789-abcd-ef0123456789
+          Link: &lt;http://example.org/foo/fcr:tx/01234567-89ab-cdef-0123-456789abcdef>; rel="[URI TBD]"
```
I think I like these changes over the previous version, where transaction handling isn't a secondary effect of a regular request (which could get weird. why make a GET just to abort a transaction?). 

Providing a URI for the transaction for interaction purposes seems good, and so it makes sense to pass the tx uri in. It also implies that a transaction is bound to a particular domain, in case we ever end up with fedora clustering. That's probably realistic. 

I guess a Link header makes sense, although I don't think it'd be unreasonable to use a custom header either.

peichman-umd replied:
Yeah, the "GET to abort a transaction" always bothered me, too.

As for the `Link` header, the rel should mean something like "this representation is this resource as viewed in the transaction identified by the link". I retained the custom `Atomic-ID` header for the requests since I didn't see anything in the standard HTTP set of headers that seemed appropriate for the meaning of "do this in the context of".

whikloj commented:
@peichman-umd I'd agree with not having a separate authorization mechanism for transactions. You **should** have WebACLs on your objects so that people can't edit them regardless of being in or outside a transaction.

peichman-umd committed:
Removed statement about transaction authorization.

Implementations should rely on WebAC to control access to resources, even in
transactions.

peichman-umd committed:
Terminology cleanup.

peichman-umd committed:
Removed transaction path-scoping.

peichman-umd committed:
Further specify committing and aborting.

bbpennel commented on index.html:
```diff
@@ -85,88 +90,200 @@
+        <dd>an HTTP resource that supports operations to create a new
+        transaction</dd>
+        <dt>transaction identifier</dt>
+        <dd>an HTTP resource that uniquely identifies a trasnaction, and which
```
Maybe just `A URI that uniquely identifies a transaction,...`?
Also there's a typo, `trasnaction`

peichman-umd replied:
Good catch.

bbpennel commented on index.html:
```diff
@@ -85,88 +90,200 @@
+          HEAD http://example.org/repo
+
+          200 OK
+          Link: &lt;http://example.org/repo/fcr:tx>; rel="[URI TBD]"
```
Looks like you have a URI for the rel further down the document now `http://fedora.info/definitions/v4/transaction#endpoint`

peichman-umd replied:
That's still tentative, and would need a (tiny) ontology document written for it, but yeah.

bbpennel commented on index.html:
```diff
@@ -85,88 +90,200 @@
+          identifier as the value of the <code>Atomic-ID</code> header, as
+          described here, until the series expires or is finished. The effect of
+          the request, if it has one, and if the request is successful, SHOULD
+          NOT be visible to clients not using that transaction identifier The
```
Missing a period `identifier The`

peichman-umd replied:
Good catch.

bbpennel commented on index.html:
```diff
@@ -85,88 +90,200 @@
-          in the series MUST become durable and MUST become visible to all clients. The unique
-          identifier of the series MUST be expired and MUST NOT be reused.
+          A client finishes an atomic series of requests by sending a
+          <code>PUT</code> request to the transaction identifier URI.
```
`PUT` to commit the transaction feels a bit unexpected to me. I guess the fcrepo4 approach of having a tx:commit subpath makes the intent clearer, but I understand that you were going for a single endpoint approach. So I'm not saying this has to change, but would be curious how others feel.

peichman-umd replied:
I was trying to come up with the most RESTful mapping of HTTP verbs to transaction actions, and `PUT` seemed to fit the best, as long as you look at it as changing the state of the transaction from "open" to "committed". `POST` felt like a better fit for extending the transaction, given the semantics we assign `POST` elsewhere in LDP for appending to a container. And I didn't want to overload `POST` with multiple, divergent meanings.

I wonder, would it be better if there were a minimal body specified, something like:
```
{"state": "committed"}
```
This could be mirrored by adding a `GET` request that would return:
```
{"state": "open"}
```
while the transaction was open.

Then we could specify that implementations are free to add their own JSON keys for implementation-specific purposes.

Thoughts?

bbpennel replied:
I don't mind this, although i can't think of anything else in fedora or its specification with this sort of interaction at the moment. There isn't really a way to see that a previous transaction has already been committed (other than being the one that commits it, which makes returning a state of "open" seem a little less useful

Now that I think about it, what happens if you include a tx id that has already been committed in an Atomic-Id header?

peichman-umd replied:
I'm inclined to say the case of using an `Atomic-ID` of an already committed transaction would return a `409 Conflict`. That would actually also apply to the case of any transaction identifier that the server didn't recognize.

peichman-umd replied:
Actually, I think that case would already be covered by this paragraph in "Adding to an atomic series":

> If the transaction identified by the <code>Atomic-ID</code> header is
> invalid for any reason, the server MUST NOT update any resources, and
> it MUST respond with a <code>409 Conflict</code> HTTP status. The
> server MAY also include more information about the reason for
> rejecting the request in the response body.

https://github.com/fcrepo/fcrepo-specification-atomic-operations/pull/8/files#diff-eacf331f0ffc35d4b482f1d15a887d3bR217-R221

whikloj replied:
I'm not sure I like linking HTTP verbs to interactions. I feel like either adding distinct `tx:commit` & `tx:rollback` or submitting some body content. But thanks just my two cents.

peichman-umd replied:
My intention was to create a RESTful interaction with a transaction resource. I agree having a body, at least for the "commit" interaction, might make that intent even clearer.

bbpennel commented on index.html:
```diff
@@ -85,88 +90,200 @@
+        </p>
+        <p>
+          The request body of this request MAY be non-empty. If it is non-empty,
+          [TODO: should we specify? or leave it up to implementations; this
```
We're not necessarily linking transactions to versions, so commit message wouldn't necessarily have anywhere to go. My inclination is to leave it up to the implementation what could go in the request body.

peichman-umd replied:
I agree.

peichman-umd committed:
Addressing feedback.

whikloj commented on index.html:
```diff
@@ -85,88 +90,203 @@
+        </p>
+        <p>
+          An implementation SHOULD permit extending the lifetime of a series by
+          a <code>POST</code> request to the transaction identifier URI.
```
If we are considering a more deliberate message to cause a `commit` to occur, then I would suggest that pattern also be used here.

peichman-umd replied:
I am open to suggestions on what that message body would look like...

bbpennel replied:
I guess I don't imagine my current application passing along any extra info during a commit, so I'd be okay with the body for POST and PUT being a MAY rather than requiring it for both

peichman-umd replied:
@whikloj I've updated my revisions to include the concept of a "commit endpoint", so clients don't need to send any request bodies to manipulate transactions.

peichman-umd committed:
Define responses to extending transactions.

peichman-umd committed:
Define a "commit endpoint"

Eliminate the use of a request body for committing a transaction in favor of a PUT to a commit endpoint. This URI of this endpoint is advertised by the server upon transaction creation via a Link header.

Also updated the example URIs to be (hopefully) clearer.

peichman-umd committed:
Servers may reject certain requests during transactions.

dbernstein commented on index.html:
```diff
@@ -85,88 +90,232 @@
-          in an atomic fashion. This is accomplished by the use of headers on requests. An atomic
-          series of operations should not require any more requests than the same series of
-          operations performed non-atomically.
+          Fedora provides factilities by which to conduct multiple operations
```
typo

dbernstein reviewed (approved):
Ignore this approval . Mistake.

dbernstein reviewed (commented):
I believe there are some modifications to some return codes after this one is merged.
